### PR TITLE
Roll fwd across hive

### PR DESCRIPTION
**Pull request by ramarag**
This change  disables user profile lookup of shared Fx
It also rolls forward across multiple probe locations

### CODE REVIEW

**Comment by ramarag** [comment]
This fixes https://github.com/dotnet/core-setup/issues/2398

@gkhanna79 @eerhardt @nguerrera @livarcocc PTAL

**Comment by ramarag** [comment]
will make the necessary changes to doc once i get a signify on the code changes

File: src/test/HostActivationTests/GivenThatICareAboutMultilevelSharedFxLookup.cs [hunk 1]
```diff
@@ -144,125 +155,164 @@ public void SharedFxLookup_Must_Verify_Folders_in_the_Correct_Order()
...
+                .HaveStdErrContaining(Path.Combine(_exeSelectedMessage, "9999.0.3"));
+
+            DeleteAvailableSharedFxVersions(_exeSharedFxBaseDir, "9999.0.0", "9999.0.0-dummy0","9999.0.3");
+            DeleteAvailableSharedFxVersions(_globalSharedFxBaseDir, "9999.0.2", "9999.0.0-dummy1", "9999.0.0-dummy2");
         }
 
         [Fact]
         public void SharedFxLookup_Must_Not_Roll_Forward_If_Framework_Version_Is_Specified_Through_Argument()
         {
+            if (RuntimeEnvironment.OperatingSystemPlatform == Platform.Windows)
```
Comment (eerhardt) [inline src/test/HostActivationTests/GivenThatICareAboutMultilevelSharedFxLookup.cs]: What is the plan here?  I don't think turning all these tests off on Windows is a great plan.

File: src/corehost/common/pal.unix.cpp [hunk 2]
```diff
@@ -204,12 +204,18 @@ bool locate_dotnet_on_path(pal::string_t* dotnet_exe)
         append_path(&tok, _X("dotnet"));
         if (pal::realpath(&tok) && is_executable(tok))
         {
-            *dotnet_exe = tok;
+            dotnet_exes->push_back(tok);
```
Comment (eerhardt) [inline src/corehost/common/pal.unix.cpp]: When are you ever returning more than one here?  You are returning 'true' as soon as you push something on the list.

Do we really need to handle more than one?

File: src/corehost/cli/fxr/fx_muxer.cpp [hunk 3]
```diff
@@ -333,27 +333,26 @@ bool fx_muxer_t::resolve_hostpolicy_dir(host_mode_t mode,
     return false;
 }
 
-void fx_muxer_t::resolve_roll_forward(pal::string_t& fx_dir,
+fx_ver_t fx_muxer_t::resolve_most_compatible(const std::vector<fx_ver_t>& version_list,
```
Comment (gkhanna79) [inline src/corehost/cli/fxr/fx_muxer.cpp]: We should cotinue to call this resolve_roll_froward since that is what this function is doing.

File: src/corehost/cli/fxr/fx_muxer.cpp [hunk 4]
```diff
@@ -488,26 +482,62 @@ pal::string_t fx_muxer_t::resolve_fx_dir(host_mode_t mode,
...
 
+        pal::string_t resolved_ver_str = resolved_ver.as_str();
+        append_path(&fx_dir, resolved_ver_str.c_str());
+
         if (pal::directory_exists(fx_dir))
         {
             trace::verbose(_X("Chose FX version [%s]"), fx_dir.c_str());
-            return fx_dir;
+
+            if (!do_roll_forward)
```
Comment (gkhanna79) [inline src/corehost/cli/fxr/fx_muxer.cpp]: This check looks incorrect since for a given dir in the hiv_dir you enumerated above, resolve_most_compatible has already performed patch roll forward if it was applicable.

Comment (gkhanna79) [inline src/corehost/cli/fxr/fx_muxer.cpp]: Are you not trying to say "if (!multi-level-lookup-enabled)"?

File: src/corehost/cli/fxr/fx_muxer.cpp [hunk 5]
```diff
@@ -488,26 +482,62 @@ pal::string_t fx_muxer_t::resolve_fx_dir(host_mode_t mode,
...
+            {
+                return fx_dir;
+            }
+
+            //Continue to search for a better match if available
+            std::vector<fx_ver_t> version_list;
```
Comment (gkhanna79) [inline src/corehost/cli/fxr/fx_muxer.cpp]: I think this version_list should be at a higher scope where you add each "most compatibly resolved" version from each hive enumerated and will not need to add selected_ver.

Once the hive enumeration is complete, you invoke resolve_most_compatible for this list and will get the most applicable version that is patch roll forwarded if applicable.

Comment (gkhanna79) [inline src/corehost/cli/fxr/fx_muxer.cpp]: With the above said, I think this will work fine as well. Please add few comments here.

File: src/corehost/common/pal.h [hunk 6]
```diff
@@ -211,7 +211,9 @@ namespace pal
     bool getenv(const char_t* name, string_t* recv);
     bool get_default_servicing_directory(string_t* recv);
     bool get_local_dotnet_dir(string_t* recv);
-    bool get_global_dotnet_dir(string_t* recv);
+    //As long as linux uses the path to find the global location of
```
Comment (gkhanna79) [inline src/corehost/common/pal.h]: I would suggest we rephrase this to say that since, on Linux, we determine global location by enuemrating the location where dotnet is present on path, there could be multiple such locations. We should also say that on Windows, it will only be one.

File: src/test/HostActivationTests/GivenThatICareAboutMultilevelSharedFxLookup.cs (hunk 1 above)
Comment (gkhanna79) [inline src/test/HostActivationTests/GivenThatICareAboutMultilevelSharedFxLookup.cs]: Agreed. @ramarag why are we disabling this test?

File: src/test/HostActivationTests/GivenThatICareAboutMultilevelSharedFxLookup.cs [hunk 7]
```diff
@@ -366,6 +414,11 @@ public void Roll_Forward_On_No_Candidate_Fx_Fails_If_No_Higher_Version_Is_Availa
         [Fact]
         public void Roll_Forward_On_No_Candidate_Fx_Must_Look_Into_All_Lookup_Folders()
         {
+            if (RuntimeEnvironment.OperatingSystemPlatform == Platform.Windows)
```
Comment (gkhanna79) [inline src/test/HostActivationTests/GivenThatICareAboutMultilevelSharedFxLookup.cs]: Same as above - this should endup doing the right thing by design. Is that not the case per your changes?

File: src/test/HostActivationTests/GivenThatICareAboutMultilevelSharedFxLookup.cs [hunk 8]
```diff
@@ -416,15 +469,21 @@ public void Roll_Forward_On_No_Candidate_Fx_Must_Look_Into_All_Lookup_Folders()
...
             // Remove dummy folders from user dir
-            DeleteAvailableSharedFxVersions(_userSharedFxBaseDir, "9999.1.1");
+            DeleteAvailableSharedFxVersions(_exeSharedFxBaseDir, "9999.1.1");
+            DeleteAvailableSharedFxVersions(_globalSharedFxBaseDir, "9999.1.0");
         }
 
         [Fact]
         public void Roll_Forward_On_No_Candidate_Fx_May_Be_Enabled_Through_Runtimeconfig_Json_File()
         {
+            if (RuntimeEnvironment.OperatingSystemPlatform == Platform.Windows)
```
Comment (gkhanna79) [inline src/test/HostActivationTests/GivenThatICareAboutMultilevelSharedFxLookup.cs]: Same as above.

File: src/test/HostActivationTests/GivenThatICareAboutMultilevelSharedFxLookup.cs [hunk 9]
```diff
@@ -461,12 +520,17 @@ public void Roll_Forward_On_No_Candidate_Fx_May_Be_Enabled_Through_Runtimeconfig
...
                 .Pass()
                 .And
-                .HaveStdErrContaining(_exeSelectedMessage);
+                .HaveStdErrContaining(_globalSelectedMessage);
         }
 
         [Fact]
         public void Roll_Forward_On_No_Candidate_Fx_May_Be_Enabled_Through_Argument()
         {
+            if (RuntimeEnvironment.OperatingSystemPlatform == Platform.Windows)
```
Comment (gkhanna79) [inline src/test/HostActivationTests/GivenThatICareAboutMultilevelSharedFxLookup.cs]: As above.

File: src/corehost/common/pal.unix.cpp (hunk 2 above)
Comment (ramarag) [inline src/corehost/common/pal.unix.cpp]: great catch ! will fix and add a test for it

File: src/test/HostActivationTests/GivenThatICareAboutMultilevelSharedFxLookup.cs (hunk 1 above)
Comment (ramarag) [inline src/test/HostActivationTests/GivenThatICareAboutMultilevelSharedFxLookup.cs]: With the removal of userprofile the only other probe location is the global location on windows, and it is not easy to mock it unless we add knobs into the host itself.

I believe once the tracking item on containarizing the tests https://github.com/dotnet/core-setup/issues/1553 is done, then we can make the infra changes necessary for enabling these tests on windows

File: src/corehost/cli/fxr/fx_muxer.cpp (hunk 4 above)
Comment (ramarag) [inline src/corehost/cli/fxr/fx_muxer.cpp]: no this is for roll_forward not enabled, will restructure the code for better readability

**Comment by ramarag** [comment]
@gkhanna79 @eerhardt Any other concerns ?

File: src/corehost/cli/fxr/fx_muxer.cpp [hunk 10]
```diff
@@ -492,22 +486,60 @@ pal::string_t fx_muxer_t::resolve_fx_dir(host_mode_t mode,
...
+            }
+
+            fx_ver_t resolved_ver = resolve_framework_version(version_list, fx_ver, specified, patch_roll_fwd_enabled, roll_fwd_on_no_candidate_fx_val);
+
+            pal::string_t resolved_ver_str = resolved_ver.as_str();
+            append_path(&fx_dir, resolved_ver_str.c_str());
+
+            if (pal::directory_exists(fx_dir))
+            {
+                trace::verbose(_X("Chose FX version [%s]"), fx_dir.c_str());
```
Comment (gkhanna79) [inline src/corehost/cli/fxr/fx_muxer.cpp]: We should display this message only for the final version that is selected. Here, however, you can add logging to display the two versions being used to determine the resolved_ver.

File: src/corehost/common/pal.h [hunk 11]
```diff
@@ -211,7 +211,9 @@ namespace pal
     bool getenv(const char_t* name, string_t* recv);
     bool get_default_servicing_directory(string_t* recv);
     bool get_local_dotnet_dir(string_t* recv);
-    bool get_global_dotnet_dir(string_t* recv);
+    //On Linux, we determine global location by enuemrating the location where dotnet is present on path, hence there could be multiple such locations
```
Comment (eerhardt) [inline src/corehost/common/pal.h]: nit: type-o `enuemrating`
